### ISSUE TITLE
Space leak with Callbacks

### ISSUE THREAD

**Issue by FPtje** [dmjio/miso#363]
# How to reproduce

1. Open the [SVG example](https://svg.haskell-miso.org/)
1. Open Chrome's task manager with shift+escape (or some other browser's equivalent)
1. Move your mouse
1. Watch memory usage go up
1. Stop moving your mouse
1. See memory usage stop going up

# Cause 
[`onWithOptions`](https://github.com/dmjio/miso/blob/master/ghcjs-src/Miso/Html/Internal.hs#L257) attributes create callbacks every time the view function is called. All callbacks are stored in a variable called `h$extraRoots`. Unless you release them with the `releaseCallback` function (see https://github.com/ghcjs/ghcjs-base/blob/master/GHCJS/Foreign/Callback.hs#L46) or its `h$release` Javascript equivalent, they're gonna pile up.

One solution is to clean them up in the Javascript side of the diffing algorithm, as that algorithm knows when callbacks can be removed.

**Comment by dmjio** [comment]
Had an open issue related to this, #72 (due to the documented behavior of space leaks if callbacks are not released). Strangely enough, the space leak behavior never occurred. It is true we are creating new `callbacks` on each redraw (any actions that cause a model diff), but the size of `h$extraRoots` always stays constant between each redraw. My only thought is that the `callbacks` are getting GC'd somehow. I'd be curious to know the real reason a space leak isn't occurring here. Maybe @luite can shed some light on this. Any ideas @luite ?

cc @basvandijk as well.

**Comment by dmjio** [comment]
The code in question is here:

https://github.com/ghcjs/ghcjs-base/blob/552da6d30fd25857a2ee8eb995d01fd4ae606d22/GHCJS/Foreign/Callback.hs#L38

**Comment by dmjio** [comment]
digging deeper here, `makeCallback` is used, defined here:

```js
// extra roots for the heap scanner: objects with root property                                                                                                        
var h$extraRootsN = 0;                                                                                                                                                 
var h$extraRoots = new h$Set();                                                                                                                                        
                                                                                                                                                                       
function h$makeCallback(f, extraArgs, action) {                                                                                                                        
    var args = extraArgs.slice(0);                                                                                                                                     
    args.unshift(action);                                                                                                                                              
    var c = function() {                                                                                                                                               
        return f.apply(this, args);                                                                                                                                    
    }                                                                                                                                                                  
    c._key = ++h$extraRootsN;                                                                                                                                          
    c.root = action;                                                                                                                                                   
    h$extraRoots.add(c);                                                                                                                                               
    return c;                                                                                                                                                          
}  
```

**Comment by luite** [comment]
can you make a small test case that you think should leak?

related:

I'm still working on the GHC 8.2 branch, making it ready for general use. For my own debugging, I've experimented with a few tools to find problems more easily. One is runtime checking of each Haskell value at each use site. This can detect malformed values (for example an `Int#` with value `3.5` or `null`) and values that are used by the RTS after being "lost" (callback not registered or freed too early).

Additional functionality is available when the JS environment supports weak refs (support in real browsers is not available, but I'm using an Electron app with a v8-specific `npm` package), and I think reporting unneeded roots would be a useful new feature (`h$extraRoots` would use weak refs in the debugging environment, and report situations where the value is already dead but still retained by `h$extraRoots`)

Unfortunately the test env requires an unreleased Electron version (build from source takes 20GB disk space) and my implementation inserting the checks is not entirely problem-free and completely kills GHCJS optimization. Hopefully I can make a more usable version available later.

**Comment by FPtje** [comment]
Hi @luite, sorry for the late reply. I can't easily reproduce what I'm seeing in Miso. I did notice something else:

```haskell
main :: IO ()
main = do
    callbacks <- forM [1..1000] $ \_ ->
      asyncCallback (pure ())

    putStrLn "Created callbacks"
    forM_ callbacks $ \cb ->
      releaseCallback cb

    putStrLn "Destroyed callbacks"
```

`h$extraRoots` contains 1000 callbacks after the creation bit, as expected, but after destroying them all, `h$extraRoots` looks like this:

![2018-03-20-192844_944x196_scrot](https://user-images.githubusercontent.com/1202014/37675202-f751944c-2c74-11e8-886a-823448139acb.png)

9 `null`s? That may be me misunderstanding how `h$ExtraRoots` works, though. I'm not well acquainted with ghcjs' callback mechanism.


Either way, I just tried reproducing this by clicking around on the https://haskell-miso.org/ website, unlike @dmjio's observation (and I could swear my own as well), `h$extraRoots` increases in size, as does `h$extraRootsN` increase in value. That's actually expected given Miso's code. Odd that we saw something different earlier. It means that we should release the callbacks as the documentation tells us to.
